pkg/apis/crd/v1alpha1/types.go

```diff
@@ -354,3 +355,128 @@ type BGPPeer struct {
+type Source struct {
+	// Pod is the source pod,
+	Pod *PodReference `json:"pod,omitempty"`
+	// IP is the source IPv4 or IPv6 address.
+	IP *string `json:"ip,omitempty"`
+}
+
+// Destination describes the destination spec of the PacketCapture.
+type Destination struct {
+	// Pod is the destination Pod, exclusive with destination Service.
+	Pod *PodReference `json:"pod,omitempty"`
+	// Service is the destination Service, exclusive with destination Pod.
+	Service *ServiceReference `json:"service,omitempty"`
+	// IP is the destination IPv4 or IPv6 address.
+	IP *string `json:"ip,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	UDP *UDPHeader `json:"udp,omitempty"`
+	TCP *TCPHeader `json:"tcp,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort *int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort *int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags *int32 `json:"flags,omitempty"`
+}
+
+// Packet includes header info.
+type Packet struct {
+	// IPFamily is the filter's IP family. Default to `IPv4`.
+	IPFamily v1.IPFamily `json:"ipFamily,omitempty"`
+	// Protocol represents the transport protocol. Default is to not filter on protocol.
+	Protocol        *intstr.IntOrString `json:"protocol,omitempty"`
+	TransportHeader TransportHeader     `json:"transportHeader"`
+}
+
+// PacketCaptureFirstNConfig contains the config for the FirstN type capture. The only supported parameter is
+// `Number` at the moment, meaning capturing the first specified number of packets in a flow.
+type PacketCaptureFirstNConfig struct {
+	Number int32 `json:"number"`
+}
+
+const DefaultPacketCaptureTimeout uint16 = 60
+
+type PacketCapturePhase string
+
+const (
+	PacketCaptureRunning   PacketCapturePhase = "Running"
+	PacketCaptureSucceeded PacketCapturePhase = "Succeeded"
+	PacketCaptureFailed    PacketCapturePhase = "Failed"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCaptureList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []PacketCapture `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCapture struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              PacketCaptureSpec   `json:"spec"`
+	Status            PacketCaptureStatus `json:"status"`
+}
+
+type CaptureConfig struct {
+	FirstN *PacketCaptureFirstNConfig `json:"firstN,omitempty"`
```

Comment:
why is it 'omitempty'? what is the default value if it can be empty?

Reply:
mainly for other capture options in the future i think.

Reply:
> I think we should keep it as a pointer, and maybe keep the omitempty, but wrap the field in a oneOf? That essentially makes the field required though, and I don't know if it's an ok way to define the API now, while giving us the opportunity to evolve the API later by adding new capture options.

This is what I would suggest too.

Reply:
@hangyan I don't think this was implemented?

But thinking about it more, I think it is actually ok to omit the `FirstN` field. If it is omitted, the capture should go on until the timeout is reached, and all captured packets should be included in the output file (although we could have an upper bound on the size of the file). That makes `FirstN` truly optional, and `oneOf` would not be appropriate. Maybe that was the intent all along. What do you think @hangyan @tnqn?

By the way this is also why we need a comment for this field.

Reply:
The the comment should probably read:

> // FirstN means we only capture first N packets from the target traffic.
> // At the moment this is the only supported configuration and it is required for every capture.

Reply:
sure, updated.